include/xtensor/xmanipulation.hpp

```diff
@@ -608,6 +608,32 @@ namespace xt
 
         return rot90_impl<n>()(std::forward<E>(e), norm_axes);
     }
+
+    template<class E>
+    inline auto roll(E&& e, std::ptrdiff_t shift) {
+      decltype(auto) cpy = empty_like(std::forward<E>(e));
+      auto eflat_view = flatten(e);
+      auto cpyflat_view = flatten(cpy);
+      if(shift == 0) {
+        std::copy(eflat_view.begin(), eflat_view.end(), cpyflat_view.begin());
+      }
+      else if(shift < 0) {
+        shift = -shift;
+        if(shift > eflat_view.size())
+          shift %= eflat_view.size();
+        std::vector<typename std::decay_t<E>::value_type> tmp{eflat_view.begin(), eflat_view.begin() + shift};
+        std::copy(tmp.begin(), tmp.end(),
+                  std::copy(eflat_view.begin() + shift, eflat_view.end(), cpyflat_view.begin()));
```

Comment:
What do you think of removing the temporary here? Since `eflat_view` and `cpy_flat_view` are distinct objects, the following should work:
```cpp
std::copy(eflat_view.begin(), eflat_view.begin() + shift,
                std::copy(eflat_view.begin() + shift, eflat_view.end(), cpyflat_view.begin()));               
```